src/adapters/morpho-aave-v2/common/morphoBasePoolAdapter.ts

```diff
@@ -0,0 +1,633 @@
+      this.provider,
+    )
+    const tokens = await this.getProtocolTokens()
+    const positionType = this.getProtocolDetails().positionType
+
+    const getBalance = async (
+      market: Erc20Metadata,
+      userAddress: string,
+      blockNumber: number,
+    ): Promise<number> => {
+      let balanceRaw
+      if (positionType === PositionType.Supply) {
+        ;[, , balanceRaw] = await lensContract.getCurrentSupplyBalanceInOf(
+          market.address,
+          userAddress,
+          { blockTag: blockNumber },
+        )
+      } else {
+        ;[, , balanceRaw] = await lensContract.getCurrentBorrowBalanceInOf(
+          market.address,
+          userAddress,
+          { blockTag: blockNumber },
+        )
+      }
+      return Number(balanceRaw)
+    }
+
+    const protocolTokensBalances = await Promise.all(
+      tokens.map(async (market) => {
+        const amount = await getBalance(market, userAddress, blockNumber!)
+        return {
+          address: market.address,
+          balance: amount,
+        }
+      }),
+    )
+
+    const protocolTokens: ProtocolPosition[] = await Promise.all(
+      protocolTokensBalances
+        .filter((protocolTokenBalance) => protocolTokenBalance.balance !== 0) // Filter out balances equal to 0
+        .map(async (protocolTokenBalance) => {
+          const balanceRaw = BigInt(protocolTokenBalance.balance)
+
+          const tokenMetadata = await this.fetchProtocolTokenMetadata(
+            protocolTokenBalance.address,
+          )
+
+          const completeTokenBalance: TokenBalance = {
+            address: protocolTokenBalance.address,
+            balanceRaw,
+            name: tokenMetadata.name,
+            symbol: tokenMetadata.symbol,
+            decimals: tokenMetadata.decimals,
+          }
+
+          const underlyingTokenBalances = await this.getUnderlyingTokenBalances(
+            {
+              userAddress,
+              protocolTokenBalance: completeTokenBalance,
+              blockNumber,
+            },
+          )
+
+          return {
+            ...protocolTokenBalance,
+            balanceRaw,
+            name: tokenMetadata.name,
+            symbol: tokenMetadata.symbol,
+            decimals: tokenMetadata.decimals,
+            type: TokenType.Protocol,
+            tokens: underlyingTokenBalances,
+          }
+        }),
+    )
+    return protocolTokens
+  }
+
+  async getWithdrawals({
+    userAddress,
+    protocolTokenAddress,
+    fromBlock,
+    toBlock,
+  }: GetEventsInput): Promise<MovementsByBlock[]> {
+    return await this.getMovements({
+      protocolToken: await this.fetchProtocolTokenMetadata(
+        protocolTokenAddress,
+      ),
+      underlyingTokens: await this.fetchUnderlyingTokensMetadata(
+        protocolTokenAddress,
+      ),
+      filter: {
+        smartContractAddress: protocolTokenAddress,
+        fromBlock,
+        toBlock,
+        from: userAddress,
+        to: undefined,
+      },
+    })
```

Comment:
Matching of a supply position is considered as a withdrawal here while, in fact, its a borrow. Why are you using pool transfers instead of just querying the withdraw events on Morpho? It seems way more easier

Reply:
You are right + it will be easier to do it on an event basis.